### PR TITLE
fix: do not swallow [Async_io] errors

### PR DESCRIPTION
**Pull request by rgrinberg**
Errors would be swallowed whenever the select thread would raise while
the mutex was unlocked. The subsequent [finally] that tries to unlock
would fail, and we'd hdie the original error. We ignore the error of
unlocking the mutex twice now.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

File: src/dune_async_io/async_io.ml [hunk 1]
```diff
@@ -224,7 +224,11 @@ let start t =
   let module Scheduler = (val t.scheduler : Scheduler) in
   Scheduler.spawn_thread (fun () ->
     Mutex.lock t.mutex;
-    Exn.protect ~f:(fun () -> select_loop t) ~finally:(fun () -> Mutex.unlock t.mutex))
+    Exn.protect
+      ~f:(fun () -> select_loop t)
+      ~finally:(fun () ->
+        try Mutex.unlock t.mutex with
```
Comment (dkalinichenko-js) [inline src/dune_async_io/async_io.ml:230]: Let's add a comment explaining why we are catching the error here.

Comment (rgrinberg) [reply src/dune_async_io/async_io.ml]: Something about this code didn't sit right with me and I rewrote it without this ugliness. Take another look.

File: src/dune_async_io/async_io.ml [hunk 2]
```diff
@@ -201,6 +206,9 @@ let rec select_loop t =
      | exception Unix.Unix_error (Unix.(EINTR | EAGAIN), _, _) ->
        Mutex.lock t.mutex;
        select_loop t
+     | exception exn ->
+       let exn = Exn_with_backtrace.capture exn in
+       raise (Unlocked exn)
```
Comment (Alizter) [inline src/dune_async_io/async_io.ml:211]: Should this be `raise_notrace`?

Comment (rgrinberg) [reply src/dune_async_io/async_io.ml]: Sure, why not. Doesn't make much of a difference.

**Review by dkalinichenko-js — APPROVED** [review]

**Pull request merged**